framework/src/main/java/org/checkerframework/framework/stub/JavaStubifier.java

```diff
@@ -0,0 +1,197 @@
+     *
+     * @param dir directory to process
+     */
+    private static void process(String dir) {
+        Path root = Paths.get(dir);
+        MinimizerCallback mc = new MinimizerCallback();
+        ProjectRoot projectRoot = new ParserCollectionStrategy().collect(root);
+
+        projectRoot
+                .getSourceRoots()
+                .forEach(
+                        sourceRoot -> {
+                            try {
+                                sourceRoot.parse("", mc);
+                            } catch (IOException e) {
+                                System.err.println("IOException: " + e);
+                            }
+                        });
+    }
+
+    /** Callback to process each Java file; see class documentation for details. */
+    private static class MinimizerCallback implements SourceRoot.Callback {
+        /** The visitor instance. */
+        private final MinimizerVisitor mv;
+
+        /** Create a MinimizerCallback instance. */
+        public MinimizerCallback() {
+            this.mv = new MinimizerVisitor();
+        }
+
+        @Override
+        public Result process(
+                Path localPath, Path absolutePath, ParseResult<CompilationUnit> result) {
+            Result res = Result.SAVE;
+            // System.out.printf("Minimizing %s%n", absolutePath);
+            Optional<CompilationUnit> opt = result.getResult();
+            if (opt.isPresent()) {
+                CompilationUnit cu = opt.get();
+                // this somehow only removes comments except the
+                // first one, and copyright headers are kept
+                cu.getComments().forEach(Node::remove);
+                mv.visit(cu, null);
+                if (cu.findAll(ClassOrInterfaceDeclaration.class).isEmpty()
+                        && cu.findAll(AnnotationDeclaration.class).isEmpty()
+                        && cu.findAll(EnumDeclaration.class).isEmpty()
+                        && !absolutePath.endsWith("package-info.java")) {
+                    // All content is removed, delete this file.
+                    new File(absolutePath.toUri()).delete();
+                    res = Result.DONT_SAVE;
+                }
+            }
+            return res;
+        }
+    }
+
+    /** Visitor to process one compilation unit; see class documentation for details. */
+    private static class MinimizerVisitor extends ModifierVisitor<Void> {
+        /** Whether to consider members implicitly public. */
+        private boolean implicitlyPublic = false;
+
+        @Override
+        public ClassOrInterfaceDeclaration visit(ClassOrInterfaceDeclaration cid, Void arg) {
+            boolean prevIP = implicitlyPublic;
+            if (cid.isInterface()) {
+                // All members of interfaces are implicitly public.
+                implicitlyPublic = true;
+            }
+            super.visit(cid, arg);
+            removeIfPrivateOrPkgPrivate(cid);
```

Comment:
This is removing `AbstractStringBuilder` because it is a package private class.  However, some of the methods are exposed in public classes.

Reply:
@mernst Thanks for noticing this!

So it looks like we can probably never remove a class. Even a private inner class can have a public subclass within the same outer class:

```java
import org.checkerframework.checker.nullness.qual.Nullable;
class Demo {
    private class Internal {
        public void foo(@Nullable Object p) {}
    }
    public class Public extends Internal {}
  
    void bar(Public p) {
        p.foo(null);
    }
}
```

The public inner class could get exposed, so we need to keep it to maintain annotations from it.

Similarly, for `AbstractStringBuilder`, even though nobody from outside the package can refer to that class, there can be visible subclasses.
I'll remove that line for classes.

We kind-of have the same problem with methods: if a package-private method uses an inheritable annotation (like `@Pure`), the method affects overriding methods that might become public.
So we should also stop removing package-private methods.
It's still fine to remove private methods, because they cannot be overridden.

Removing private and package-private fields still seems ok to me.

Any other cases I'm missing?

Reply:
That seems like a reasonable choice for now.  If the stubs remain too big (or if we have accidentally left something out), we can revisit the design later.